lib/mapObjects/CRewardableObject.cpp

```diff
@@ -359,8 +362,16 @@ void CRewardableObject::grantRewardAfterLevelup(const CRewardVisitInfo & info, c
 
 		cb->giveCreatures(this, hero, creatures, false);
 	}
-
-	if(info.reward.removeObject)
+	
+	if(!info.reward.casts.empty())
+	{
+		caster = std::make_unique<spells::OuterCaster>(hero, SecSkillLevel::EXPERT);
+		for(const auto & c : info.reward.casts)
+		{
+			cb->castSpell(caster.get(), c, int3{-1, -1, -1});
```

Comment:
Perhaps we should add some handling is spellcasting is impossible?
- No owned towns for TP
- All owned towns are blocked by visitor for TP
- No water tiles nearby for Summon Boat
- Movement check for both DD and TP

Will this display message understandable to player?
If not - we might want to have some handling for this scenario.

Reply:
But all those checks are inside spell mechanics. I don't want to introduce any spell-specific logic here

Reply:
Maybe. But visiting "Town Gate" object only to see message like "Failed to cast a spell" is weird and not what player expects.
Even basic
```
if (cb->canCast())
     cb->castSpell()
else
    showMessage("You approach Town Gate, which were supposed to bring you back to your town, but for some reason gates stay closed. ")  // <- message configured in reward definition
```
would be better

Frankly I would check HotA first, at the very least they already faced these issues (how to handle non-functional gates) and might already have some handling for it.

Reply:
> Frankly I would check HotA first, at the very least they already faced these issues (how to handle non-functional gates) and might already have some handling for it.

It's not applicable for hota, because in their case town gate works always. 
In our case we may have failures like in example with boat summoning - hota just doesn't have such object.

All this can-spell-checking logic are deeply inside spell mechanics, so I don't want to deal with that (I was very surprised that it's not possible event to cast spell from lib code - only from server)

As for messages - they are quite meaningful for spells. You won't see "Failed to cast spell", most probably "Hero is tired - try again tomorrow"

I don't want to make silver-bullet changes covering all possible use-cases. Yes - for now it works with limitations, but it solves well very specific problem of supporting town gate from hota mod.

Reply:
Well, this is hota with 0 towns.
So our behavior shoud be as good as hota then 😅 
![зображення](https://user-images.githubusercontent.com/1576820/230888364-04a68b97-4670-4ff3-9f04-a5357ec8798e.png)

Reply:
are we able to easily fix taking movement points though?